### ISSUE TITLE
Adding folder using Long UNC path causes stack overflow

### ISSUE THREAD

**Issue by ProactiveServices** [syncthing/syncthing#3577]
Syncthing Version: v0.14.6
OS Version: Windows 7 x64

When adding a folder using a \\\\?\ -prefixed UNC path Syncthing crashes with a stack overflow. Syncthing does seem to support \\\\?\, as manually editing the config file and using the prefix results in a working sync.

Steps to reproduce:
Click "Add folder"
In folder path type: \\\\?\\C:\Users
Syncthing should crash as you're typing/just afterwards.

[stracktrace.txt](https://github.com/syncthing/syncthing/files/461210/stracktrace.txt)

**Comment by wweich** [comment]
AFAIK, UNC is `\\machine\...` not `\?\...`.
IIRC Syncthing (on windows) automatically converts local paths to `\?\...` to support paths longer than 255 characters. So there is no need to enter it manually.

**Comment by AudriusButkevicius** [comment]
Also, it should be \\?\, yet it's a bug anyway, as it shouldn't crash to start with.

**Comment by ProactiveServices** [comment]
Github ate my backspaces :-)
Updated OP appropriately.

UNC has various conventions, one of which is \\\\machine\\ (the most common) but there is also \\\\?\\ which asks Windows not to do any string processing and permits long file names, and a few other conventions.

**Comment by ProactiveServices** [comment]
@wweich: Syncthing doesn't automatically convert such paths - when using a non-UNC path, you get "file name too long" errors.

**Comment by AudriusButkevicius** [comment]
So syncthing actually wedges in the UNC prefix on Windows automatically, so perhaps the fact that you essentially wedge it in twice is causing the crash?

**Comment by ProactiveServices** [comment]
If I edit config.xml and change C:\Folder to \\\\?\C:\Folder, Syncthing works without a problem. I've had to do this to a few shared folders because they have deeply-nested paths which weren't working when the folder path was C:\Folder.

**Comment by ProactiveServices** [comment]
Forgot steps to reproduce, sorry. Added to OP.

**Comment by AudriusButkevicius** [comment]
Well syncthing adds \\?\ by itself, so not sure why you are adding it yourself.

**Comment by calmh** [comment]
So what I think happens here is that the path finally passed to glob becomes `\\?\\?\C:\Users` which it doesn't like. I tried reproducing it with a few hacks on Mac but couldn't so leaving for some Windows developer.

**Comment by ProactiveServices** [comment]
I swear I came across a recent instance of Syncthing choking on long file names but can't seem to re-create it, so it seems it's just a GUI bug at play.

**Comment by nrm21** [comment]
Tried to look into this a bit but couldn't figure it out.  Knowing almost nothing of Angular and JavaScript doesn't help much either, but I tried to stackexchange and google my way through that.  Here's what I got so far:

In syncthingController.js there is a $scope.$watch() function that reads a 'newvalue' and calls a rest call using it as a parameter every single time someone types a letter in the textbox.  As soon as someone types the colon after the '\\?\C' is when the crash occurs (actually a very short delay of half a second or so after to be precise).

Maybe someone good with JavaScript can figure out how to remove any \\?\ input before the path since it isn't necessary.
